action::stable_sort of vector broken on Clang 3.8.1 since ~last Xmas

This code used to compile and run under Clang 3.8.1:

~~~~~.cpp
#include <range/v3/all.hpp>

#include <iostream>
#include <vector>

int main() {
	const std::vector<double> scores = { 3.0, 1.0, 2.0 };
	std::vector<int> indices = { 0, 1, 2 };

	indices |= ranges::action::stable_sort(
		ranges::less{},
		[&] (const int &x) { return scores[ x ]; }
	);

	std::cerr << ( indices | ranges::view::all ) << "\n";
}
~~~~~

...but since 61917222962c9cb5e1056bdc175cf31ab1894f1c, it gives errors:

~~~~~
In file included from range-v3.stable-sort-prob.cpp:1:
In file included from range-v3/include/range/v3/all.hpp:19:
In file included from range-v3/include/range/v3/action.hpp:30:
range-v3/include/range/v3/action/stable_sort.hpp:73:21: error: static_assert failed "The object on which action::stable_sort operates must be a model of the ForwardRange concept."
                    CONCEPT_ASSERT_MSG(ForwardRange<Rng>(),
                    ^                  ~~~~~~~~~~~~~~~~~~~
range-v3/include/range/v3/utility/concepts.hpp:812:28: note: expanded from macro 'CONCEPT_ASSERT_MSG'
#define CONCEPT_ASSERT_MSG static_assert
                           ^
range-v3/include/range/v3/utility/invoke.hpp:124:17: note: in instantiation of function template specialization 'ranges::v3::action::stable_sort_fn::operator()<ranges::v3::less &, (lambda at range-v3.stable-sort-prob.cpp:12:3), ranges::v3::ident, 42, 0>' requested here
                detail::forward<F>(fn)(detail::forward<Args>(args)...)
                ^
In file included from range-v3.stable-sort-prob.cpp:1:
In file included from range-v3/include/range/v3/all.hpp:17:
In file included from range-v3/include/range/v3/core.hpp:18:
In file included from range-v3/include/range/v3/distance.hpp:22:
In file included from range-v3/include/range/v3/range_traits.hpp:24:
range-v3/include/range/v3/range_concepts.hpp:64:45: error: no matching function for call to object of type 'const ranges::v3::adl_begin_end_detail::begin_fn'
                using iterator_t = decltype(begin(std::declval<T&>()));
                                            ^~~~~
range-v3/include/range/v3/range_traits.hpp:35:9: note: in instantiation of template type alias 'iterator_t' requested here
        using range_iterator_t = concepts::Range::iterator_t<Rng>;
        ^
range-v3/include/range/v3/action/stable_sort.hpp:76:31: note: in instantiation of template type alias 'range_iterator_t' requested here
                    using I = range_iterator_t<Rng>;
                              ^
range-v3/include/range/v3/utility/invoke.hpp:124:17: note: in instantiation of function template specialization 'ranges::v3::action::stable_sort_fn::operator()<ranges::v3::less &, (lambda at range-v3.stable-sort-prob.cpp:12:3), ranges::v3::ident, 42, 0>' requested here
                detail::forward<F>(fn)(detail::forward<Args>(args)...)
                ^
range-v3/include/range/v3/begin_end.hpp:105:32: note: candidate template ignored: substitution failure [with Rng = ranges::v3::less &]: no matching function for call to 'impl'
                constexpr auto operator()(Rng && rng) const
                               ^
In file included from range-v3.stable-sort-prob.cpp:1:
In file included from range-v3/include/range/v3/all.hpp:19:
In file included from range-v3/include/range/v3/action.hpp:30:
range-v3/include/range/v3/action/stable_sort.hpp:77:21: error: static_assert failed "The projection function must accept objects of the iterator's value type, reference type, and common reference type."
                    CONCEPT_ASSERT_MSG(IndirectInvocable<P, I>(),
                    ^                  ~~~~~~~~~~~~~~~~~~~~~~~~~
range-v3/include/range/v3/utility/concepts.hpp:812:28: note: expanded from macro 'CONCEPT_ASSERT_MSG'
#define CONCEPT_ASSERT_MSG static_assert
                           ^
range-v3/include/range/v3/utility/invoke.hpp:124:17: note: in instantiation of function template specialization 'ranges::v3::action::stable_sort_fn::operator()<ranges::v3::less &, (lambda at range-v3.stable-sort-prob.cpp:12:3), ranges::v3::ident, 42, 0>' requested here
                detail::forward<F>(fn)(detail::forward<Args>(args)...)
                ^
In file included from range-v3.stable-sort-prob.cpp:1:
In file included from range-v3/include/range/v3/all.hpp:17:
In file included from range-v3/include/range/v3/core.hpp:17:
In file included from range-v3/include/range/v3/begin_end.hpp:21:
In file included from range-v3/include/range/v3/range_fwd.hpp:20:
range-v3/include/meta/meta.hpp:140:9: error: no type named 'type' in 'meta::v1::detail::_if_<meta::v1::list<std::__1::integral_constant<bool, false>, ranges::v3::detail::projected_<int, ranges::v3::ident> >, bool>'
        using _t = typename T::type;
        ^~~~~
range-v3/include/meta/meta.hpp:1110:9: note: in instantiation of template type alias '_t' requested here
        using if_c = _t<detail::_if_<list<bool_<If>, Args...>>>;
        ^
range-v3/include/range/v3/utility/iterator_concepts.hpp:568:9: note: in instantiation of template type alias 'if_c' requested here
        using projected = meta::if_c<IndirectInvocable<Proj, I>::value, detail::projected_<I, Proj>>;
        ^
range-v3/include/range/v3/action/stable_sort.hpp:80:60: note: in instantiation of template type alias 'projected' requested here
                    CONCEPT_ASSERT_MSG(IndirectRelation<C, projected<I, P>>(),
                                                           ^
range-v3/include/range/v3/utility/invoke.hpp:124:17: note: in instantiation of function template specialization 'ranges::v3::action::stable_sort_fn::operator()<ranges::v3::less &, (lambda at range-v3.stable-sort-prob.cpp:12:3), ranges::v3::ident, 42, 0>' requested here
                detail::forward<F>(fn)(detail::forward<Args>(args)...)
                ^
4 errors generated.
~~~~~

From what I can see on Wandbox, the problem disappears for Clang &ge; 4.0.

--- a/include/meta/meta.hpp
+++ b/include/meta/meta.hpp
@@ -1125,31 +1125,38 @@ namespace meta
         /// \cond
         namespace detail
         {
-            template <typename... Bools>
-            struct _and_;
+            template <bool>
+            struct _and_
+            {
+                template <class = void>
+                using invoke = std::true_type;
+            };
 
             template <>
-            struct _and_<> : std::true_type
+            struct _and_<false>
             {
+                template <typename Bool_, typename... Bools>
+                using invoke = invoke<
+                    if_c<!Bool_::type::value,
+                         id<std::false_type>,
+                         _and_<0==sizeof...(Bools)>>, Bools...>;
             };
 
-            template <typename Bool_, typename... Bools>
-            struct _and_<Bool_, Bools...>
-                : if_c<!Bool_::type::value, std::false_type, _and_<Bools...>>
+            template <bool>
+            struct _or_
             {
+                template <class = void>
+                using invoke = std::false_type;
             };
-
-            template <typename... Bools>
-            struct _or_;
 
             template <>
-            struct _or_<> : std::false_type
+            struct _or_<false>
             {
-            };
-
-            template <typename Bool_, typename... Bools>
-            struct _or_<Bool_, Bools...> : if_c<Bool_::type::value, std::true_type, _or_<Bools...>>
-            {
+                template <typename Bool_, typename... Bools>
+                using invoke = invoke<
+                    if_c<Bool_::type::value,
+                         id<std::true_type>,
+                         _or_<0 == sizeof...(Bools)>>, Bools...>;
             };
         } // namespace detail
         /// \endcond
@@ -1189,8 +1196,9 @@ namespace meta
         /// \e with
         /// short-circuiting.
         /// \ingroup logical
+        // Make a trip through defer<> to avoid CWG1430
         template <typename... Bools>
-        using and_ = _t<detail::_and_<Bools...>>;
+        using and_ = _t<defer<detail::_and_<0 == sizeof...(Bools)>::template invoke, Bools...>>;
 
         /// Logically or together all the Boolean parameters
         /// \ingroup logical
@@ -1209,8 +1217,9 @@ namespace meta
         /// \e with
         /// short-circuiting.
         /// \ingroup logical
+        // Make a trip through defer<> to avoid CWG1430
         template <typename... Bools>
-        using or_ = _t<detail::_or_<Bools...>>;
+        using or_ = _t<defer<detail::_or_<0 == sizeof...(Bools)>:: template invoke, Bools...>>;
 
         namespace lazy
         {

--- a/include/range/v3/action/action.hpp
+++ b/include/range/v3/action/action.hpp
@@ -63,11 +63,18 @@ namespace ranges
             private:
                 Action action_;
                 friend pipeable_access;
+
+                template<typename Rng, typename...Rest>
+                using ActionConcept = meta::and_<
+                    Range<Rng>,
+                    Invocable<Action const&, Rng, Rest...>>;
+
                 template<typename Rng>
                 using ActionPipeConcept = meta::and_<
                     Invocable<Action&, Rng>,
                     Range<Rng>,
                     meta::not_<std::is_reference<Rng>>>;
+
                 // Pipeing requires things are passed by value.
                 template<typename Rng, typename Act,
                     CONCEPT_REQUIRES_(ActionPipeConcept<Rng>())>
@@ -76,6 +83,7 @@ namespace ranges
                 (
                     invoke(act.action_, detail::move(rng))
                 )
+
             #ifndef RANGES_DOXYGEN_INVOKED
                 // For better error messages:
                 template<typename Rng, typename Act,
@@ -94,19 +102,22 @@ namespace ranges
                         "Or, wrap the argument with std::ref to pass it by reference.");
                 }
             #endif
+
             public:
                 action() = default;
                 action(Action a)
                   : action_(detail::move(a))
                 {}
+
                 // Calling directly requires things are passed by reference.
                 template<typename Rng, typename...Rest,
-                    CONCEPT_REQUIRES_(Range<Rng &>() && Invocable<Action const&, Rng &, Rest...>())>
+                    CONCEPT_REQUIRES_(ActionConcept<Rng &, Rest...>())>
                 auto operator()(Rng & rng, Rest &&... rest) const
                 RANGES_DECLTYPE_AUTO_RETURN
                 (
                     invoke(action_, rng, static_cast<Rest&&>(rest)...)
                 )
+
                 // Currying overload.
                 template<typename T, typename...Rest, typename A = Action>
                 auto operator()(T && t, Rest &&... rest) const

--- a/include/range/v3/action/join.hpp
+++ b/include/range/v3/action/join.hpp
@@ -43,19 +43,24 @@ namespace ranges
                         (bool) ranges::Container<range_value_type_t<Rng>>(),
                         range_value_type_t<Rng>,
                         std::vector<range_value_type_t<range_value_type_t<Rng>>>>;
+
+                struct Joinable_
+                {
+                    template<typename Rng>
+                    auto requires_() -> decltype(
+                        concepts::valid_expr(
+                            concepts::model_of<concepts::InputRange, Rng>(),
+                            concepts::model_of<concepts::InputRange, range_value_type_t<Rng>>(),
+                            concepts::model_of<concepts::SemiRegular, join_value_t<Rng>>()
+                        ));
+                };
+
             public:
                 template<typename Rng>
-                using Concept = meta::and_<
-                    InputRange<Rng>,
-                    meta::lazy::invoke<
-                        meta::compose<meta::quote<InputRange>, meta::quote<range_value_type_t>>,
-                        Rng>,
-                    meta::lazy::invoke<
-                        meta::compose<meta::quote<SemiRegular>, meta::quote<join_value_t>>,
-                        Rng>>;
+                using Joinable = concepts::models<Joinable_, Rng>;
 
                 template<typename Rng,
-                    CONCEPT_REQUIRES_(Concept<Rng>())>
+                    CONCEPT_REQUIRES_(Joinable<Rng>())>
                 join_value_t<Rng> operator()(Rng && rng) const
                 {
                     join_value_t<Rng> ret;
@@ -66,9 +71,9 @@ namespace ranges
                 }
 
             #ifndef RANGES_DOXYGEN_INVOKED
-                template<typename Rng, typename T,
-                    CONCEPT_REQUIRES_(!Concept<Rng>())>
-                void operator()(Rng &&, T &&) const
+                template<typename Rng,
+                    CONCEPT_REQUIRES_(!Joinable<Rng>())>
+                void operator()(Rng &&) const
                 {
                     CONCEPT_ASSERT_MSG(InputRange<Rng>(),
                         "The object on which action::join operates must be a model of the "

--- a/include/range/v3/action/join.hpp
+++ b/include/range/v3/action/join.hpp
@@ -47,8 +47,12 @@ namespace ranges
                 template<typename Rng>
                 using Concept = meta::and_<
                     InputRange<Rng>,
-                    InputRange<range_value_type_t<Rng>>,
-                    SemiRegular<join_value_t<Rng>>>;
+                    meta::lazy::invoke<
+                        meta::compose<meta::quote<InputRange>, meta::quote<range_value_type_t>>,
+                        Rng>,
+                    meta::lazy::invoke<
+                        meta::compose<meta::quote<SemiRegular>, meta::quote<join_value_t>>,
+                        Rng>>;
 
                 template<typename Rng,
                     CONCEPT_REQUIRES_(Concept<Rng>())>

--- a/include/range/v3/action/reverse.hpp
+++ b/include/range/v3/action/reverse.hpp
@@ -30,30 +30,27 @@ namespace ranges
         /// @{
         namespace action
         {
-            /// \cond
-            namespace reverse_detail
-            {
-                template<typename Rng>
-                using Concept = meta::and_<
-                    BidirectionalRange<Rng>,
-                    // Only evaluate this one if the previous one succeeded
-                    meta::lazy::invoke<
-                        meta::compose<
-                            meta::quote<Permutable>,
-                            meta::quote<iterator_t>>,
-                        Rng>>;
-            }  // namespace reverse_detail
-            /// \endcond
-
             /// Reversed the source range in-place.
             struct reverse_fn
             {
             private:
                 friend action_access;
 
+                struct Reversible_
+                {
+                    template<typename Rng>
+                    auto requires_() -> decltype(
+                        concepts::valid_expr(
+                            concepts::model_of<concepts::BidirectionalRange, Rng>(),
+                            concepts::is_true(Permutable<iterator_t<Rng>>())
+                        ));
+                };
             public:
+                template<typename Rng>
+                using Reversible = concepts::models<Reversible_, Rng>;
+
                 template<typename Rng,
-                    CONCEPT_REQUIRES_(reverse_detail::Concept<Rng>())>
+                    CONCEPT_REQUIRES_(Reversible<Rng>())>
                 Rng operator()(Rng && rng) const
                 {
                     ranges::reverse(rng);
@@ -62,7 +59,7 @@ namespace ranges
 
             #ifndef RANGES_DOXYGEN_INVOKED
                 template<typename Rng,
-                         CONCEPT_REQUIRES_(!reverse_detail::Concept<Rng>())>
+                         CONCEPT_REQUIRES_(!Reversible<Rng>())>
                 void operator()(Rng &&) const
                 {
                     CONCEPT_ASSERT_MSG(BidirectionalRange<Rng>(),

--- a/include/range/v3/action/reverse.hpp
+++ b/include/range/v3/action/reverse.hpp
@@ -52,8 +52,8 @@ namespace ranges
                 friend action_access;
 
             public:
-                template<typename Rng, typename I = iterator_t<Rng>,
-                    CONCEPT_REQUIRES_(BidirectionalRange<Rng>() && Permutable<I>())>
+                template<typename Rng,
+                    CONCEPT_REQUIRES_(reverse_detail::Concept<Rng>())>
                 Rng operator()(Rng && rng) const
                 {
                     ranges::reverse(rng);

--- a/include/range/v3/action/sort.hpp
+++ b/include/range/v3/action/sort.hpp
@@ -45,7 +45,12 @@ namespace ranges
             public:
                 template<typename Rng, typename C = ordered_less, typename P = ident>
                 using Concept = meta::and_<
-                    ForwardRange<Rng>, Sortable<iterator_t<Rng>, C, P>>;
+                    ForwardRange<Rng>,
+                    meta::lazy::invoke<
+                        meta::compose<
+                            meta::bind_back<meta::quote<Sortable>, C, P>,
+                            meta::quote<iterator_t>>,
+                        Rng>>;
 
                 template<typename Rng, typename C = ordered_less, typename P = ident,
                     CONCEPT_REQUIRES_(Concept<Rng, C, P>())>

--- a/include/range/v3/action/sort.hpp
+++ b/include/range/v3/action/sort.hpp
@@ -42,18 +42,24 @@ namespace ranges
                     std::bind(sort, std::placeholders::_1, protect(std::move(pred)),
                         protect(std::move(proj)))
                 )
+
+                struct Sortable_
+                {
+                    template<typename Rng, typename C = ordered_less, typename P = ident,
+                        typename I = iterator_t<Rng>>
+                    auto requires_() -> decltype(
+                        concepts::valid_expr(
+                            concepts::model_of<concepts::ForwardRange, Rng>(),
+                            concepts::is_true(ranges::Sortable<I, C, P>())
+                        ));
+                };
+
             public:
                 template<typename Rng, typename C = ordered_less, typename P = ident>
-                using Concept = meta::and_<
-                    ForwardRange<Rng>,
-                    meta::lazy::invoke<
-                        meta::compose<
-                            meta::bind_back<meta::quote<Sortable>, C, P>,
-                            meta::quote<iterator_t>>,
-                        Rng>>;
+                using Sortable = concepts::models<Sortable_, Rng, C, P>;
 
                 template<typename Rng, typename C = ordered_less, typename P = ident,
-                    CONCEPT_REQUIRES_(Concept<Rng, C, P>())>
+                    CONCEPT_REQUIRES_(Sortable<Rng, C, P>())>
                 Rng operator()(Rng && rng, C pred = C{}, P proj = P{}) const
                 {
                     ranges::sort(rng, std::move(pred), std::move(proj));
@@ -62,7 +68,7 @@ namespace ranges
 
             #ifndef RANGES_DOXYGEN_INVOKED
                 template<typename Rng, typename C = ordered_less, typename P = ident,
-                    CONCEPT_REQUIRES_(!Concept<Rng, C, P>())>
+                    CONCEPT_REQUIRES_(!Sortable<Rng, C, P>())>
                 void operator()(Rng &&, C && = C{}, P && = P{}) const
                 {
                     CONCEPT_ASSERT_MSG(ForwardRange<Rng>(),

--- a/include/range/v3/action/stable_sort.hpp
+++ b/include/range/v3/action/stable_sort.hpp
@@ -17,6 +17,7 @@
 #include <functional>
 #include <range/v3/range_fwd.hpp>
 #include <range/v3/action/action.hpp>
+#include <range/v3/action/sort.hpp>
 #include <range/v3/algorithm/stable_sort.hpp>
 #include <range/v3/utility/functional.hpp>
 #include <range/v3/utility/iterator_concepts.hpp>
@@ -43,17 +44,8 @@ namespace ranges
                         protect(std::move(proj)))
                 )
             public:
-                template<typename Rng, typename C = ordered_less, typename P = ident>
-                using Concept = meta::and_<
-                    ForwardRange<Rng>,
-                    meta::lazy::invoke<
-                        meta::compose<
-                            meta::bind_back<meta::quote<Sortable>, C, P>,
-                            meta::quote<iterator_t>>,
-                        Rng>>;
-
                 template<typename Rng, typename C = ordered_less, typename P = ident,
-                    CONCEPT_REQUIRES_(Concept<Rng, C, P>())>
+                    CONCEPT_REQUIRES_(sort_fn::Sortable<Rng, C, P>())>
                 Rng operator()(Rng && rng, C pred = C{}, P proj = P{}) const
                 {
                     ranges::stable_sort(rng, std::move(pred), std::move(proj));
@@ -62,7 +54,7 @@ namespace ranges
 
             #ifndef RANGES_DOXYGEN_INVOKED
                 template<typename Rng, typename C = ordered_less, typename P = ident,
-                    CONCEPT_REQUIRES_(!Concept<Rng, C, P>())>
+                    CONCEPT_REQUIRES_(!sort_fn::Sortable<Rng, C, P>())>
                 void operator()(Rng &&, C && = C{}, P && = P{}) const
                 {
                     CONCEPT_ASSERT_MSG(ForwardRange<Rng>(),

--- a/include/range/v3/action/stable_sort.hpp
+++ b/include/range/v3/action/stable_sort.hpp
@@ -43,19 +43,14 @@ namespace ranges
                         protect(std::move(proj)))
                 )
             public:
-                struct ConceptImpl
-                {
-                    template<typename Rng, typename C = ordered_less, typename P = ident,
-                        typename I = iterator_t<Rng>>
-                    auto requires_() -> decltype(
-                        concepts::valid_expr(
-                            concepts::model_of<concepts::ForwardRange, Rng>(),
-                            concepts::is_true(Sortable<I, C, P>())
-                        ));
-                };
-
                 template<typename Rng, typename C = ordered_less, typename P = ident>
-                using Concept = concepts::models<ConceptImpl, Rng, C, P>;
+                using Concept = meta::and_<
+                    ForwardRange<Rng>,
+                    meta::lazy::invoke<
+                        meta::compose<
+                            meta::bind_back<meta::quote<Sortable>, C, P>,
+                            meta::quote<iterator_t>>,
+                        Rng>>;
 
                 template<typename Rng, typename C = ordered_less, typename P = ident,
                     CONCEPT_REQUIRES_(Concept<Rng, C, P>())>

--- a/include/range/v3/range_fwd.hpp
+++ b/include/range/v3/range_fwd.hpp
@@ -258,37 +258,73 @@ namespace ranges
                 return 42;
             }
 
-            #if defined(__clang__) && !defined(_LIBCPP_VERSION)
-                template<typename T, typename Arg = T>
-                struct is_trivially_copy_assignable
-                  : meta::bool_<__is_trivially_assignable(T &, Arg const&)>
-                {};
-                template<typename T, typename Arg = T>
-                struct is_trivially_move_assignable
-                  : meta::bool_<__is_trivially_assignable(T &, Arg &&)>
-                {};
-             #elif defined(__GNUC__) && !defined(__clang__) && __GNUC__ < 5
-                template<typename T>
-                using is_trivially_copy_assignable = std::is_trivial<T>;
+            template<int I>
+            struct priority_tag
+              : priority_tag<I - 1>
+            {};
 
-                template<typename T>
-                using is_trivially_move_assignable = std::is_trivial<T>;
-            #else
-                using std::is_trivially_copy_assignable;
-                using std::is_trivially_move_assignable;
-            #endif
+            template<>
+            struct priority_tag<0>
+            {};
 
-            #if RANGES_CXX_LIB_IS_FINAL > 0
-                #if defined(__clang__) && !defined(_LIBCPP_VERSION)
-                    template<typename T>
-                    using is_final = meta::bool_<__is_final(T)>;
-                #else
-                    using std::is_final;
-                #endif
-            #else
-                template<typename T>
-                using is_final = std::false_type;
-            #endif
+        #if defined(__clang__) && !defined(_LIBCPP_VERSION)
+            template<typename T, typename Arg = T>
+            struct is_trivially_copy_assignable
+                : meta::bool_<__is_trivially_assignable(T &, Arg const&)>
+            {};
+            template<typename T, typename Arg = T>
+            struct is_trivially_move_assignable
+                : meta::bool_<__is_trivially_assignable(T &, Arg &&)>
+            {};
+        #elif defined(__GNUC__) && !defined(__clang__) && __GNUC__ < 5
+            template<typename T>
+            using is_trivially_copy_assignable = std::is_trivial<T>;
+
+            template<typename T>
+            using is_trivially_move_assignable = std::is_trivial<T>;
+        #else
+            using std::is_trivially_copy_assignable;
+            using std::is_trivially_move_assignable;
+        #endif
+
+        #if RANGES_CXX_LIB_IS_FINAL > 0
+        # if defined(__clang__) && !defined(_LIBCPP_VERSION)
+            template<typename T>
+            using is_final = meta::bool_<__is_final(T)>;
+        # else
+            using std::is_final;
+        # endif
+        #else
+            template<typename T>
+            using is_final = std::false_type;
+        #endif
+
+            // Work around libc++'s buggy std::is_function
+        #if !defined(_LIBCPP_VERSION) || _LIBCPP_VERSION >= 3800
+            using std::is_function;
+        #else
+            // Function types here:
+            template<typename T>
+            char (&is_function_impl_(priority_tag<0>))[1];
+
+            // Array types here:
+            template<typename T, typename = decltype((*(T*)0)[0])>
+            char (&is_function_impl_(priority_tag<1>))[2];
+
+            // Anything that can be returned from a function here (including
+            // void and reference types):
+            template<typename T, typename = T(*)()>
+            char (&is_function_impl_(priority_tag<2>))[3];
+
+            // Classes and unions (including abstract types) here:
+            template<typename T, typename = int T::*>
+            char (&is_function_impl_(priority_tag<3>))[4];
+
+            template <typename T>
+            struct is_function
+              : meta::bool_<sizeof(detail::is_function_impl_<T>(priority_tag<3>{})) == 1>
+            {};
+        #endif
 
             template<typename T>
             struct remove_rvalue_reference
@@ -304,10 +340,6 @@ namespace ranges
 
             template<typename T>
             using remove_rvalue_reference_t = meta::_t<remove_rvalue_reference<T>>;
-
-            template<int I>
-            struct priority_tag : priority_tag<I - 1> {};
-            template<> struct priority_tag<0> {};
         }
         /// \endcond
 

--- a/include/range/v3/utility/functional.hpp
+++ b/include/range/v3/utility/functional.hpp
@@ -450,13 +450,7 @@ namespace ranges
             using base_t::first;
             using base_t::second;
         public:
-            CONCEPT_REQUIRES(meta::and_<
-                DefaultConstructible<First>, DefaultConstructible<Rest>...>())
-            constexpr overloaded()
-                noexcept(meta::strict_and<
-                    std::is_nothrow_default_constructible<First>,
-                    std::is_nothrow_default_constructible<Rest>...>::value)
-            {}
+            overloaded() = default;
             constexpr overloaded(First first, Rest... rest)
               : overloaded::compressed_pair{
                     detail::move(first),

--- a/include/range/v3/utility/invoke.hpp
+++ b/include/range/v3/utility/invoke.hpp
@@ -102,7 +102,7 @@ namespace ranges
             )
         public:
             template<typename F, typename Obj, typename First, typename... Rest,
-                meta::if_c<std::is_function<F>::value, int> = 0>
+                meta::if_c<detail::is_function<F>::value, int> = 0>
             constexpr auto operator()(F (Obj::*ptr), First && first, Rest &&... rest) const
             RANGES_DECLTYPE_AUTO_RETURN_NOEXCEPT
             (
@@ -111,7 +111,7 @@ namespace ranges
                     ptr, static_cast<First&&>(first), static_cast<Rest&&>(rest)...)
             )
             template<typename Data, typename Obj, typename First,
-                meta::if_c<!std::is_function<Data>::value, int> = 0>
+                meta::if_c<!detail::is_function<Data>::value, int> = 0>
             constexpr auto operator()(Data (Obj::*ptr), First && first) const
             RANGES_DECLTYPE_AUTO_RETURN_NOEXCEPT
             (

--- a/include/range/v3/view/for_each.hpp
+++ b/include/range/v3/view/for_each.hpp
@@ -47,16 +47,21 @@ namespace ranges
                 (
                     make_pipeable(std::bind(for_each, std::placeholders::_1, protect(std::move(fun))))
                 )
+
+                struct Concept_
+                {
+                    template<typename Rng, typename Fun>
+                    auto requires_() -> decltype(
+                        concepts::valid_expr(
+                            concepts::is_true(transform_fn::Concept<Rng, Fun>()),
+                            concepts::model_of<
+                                concepts::InputRange,
+                                concepts::Invocable::result_t<Fun&, range_reference_t<Rng>>>()));
+                };
+
             public:
                 template<typename Rng, typename Fun>
-                using Concept = meta::and_<
-                    transform_fn::Concept<Rng, Fun>,
-                    meta::lazy::invoke<
-                        meta::compose<
-                            meta::quote<InputRange>,
-                            meta::bind_front<meta::quote<concepts::Invocable::result_t>, Fun&>,
-                            meta::quote<range_reference_t>>,
-                        Rng>>;
+                using Concept = concepts::models<Concept_, Rng, Fun>;
 
                 template<typename Rng, typename Fun,
                     CONCEPT_REQUIRES_(Concept<Rng, Fun>())>

--- a/include/range/v3/view/view.hpp
+++ b/include/range/v3/view/view.hpp
@@ -93,6 +93,7 @@ namespace ranges
                 (
                     v.view_(static_cast<Rng&&>(rng))
                 )
+
             #ifndef RANGES_DOXYGEN_INVOKED
                 // For better error messages:
                 template<typename Rng, typename Vw,
@@ -110,11 +111,13 @@ namespace ranges
                         "a named variable, and then pipe it to the view.");
                 }
             #endif
+
             public:
                 view() = default;
                 view(View a)
                   : view_(std::move(a))
                 {}
+
                 // Calling directly requires View arguments or lvalue containers.
                 template<typename Rng, typename...Rest,
                     CONCEPT_REQUIRES_(ViewConcept<Rng, Rest...>())>
@@ -123,6 +126,7 @@ namespace ranges
                 (
                     view_(static_cast<Rng&&>(rng), static_cast<Rest&&>(rest)...)
                 )
+
                 // Currying overload.
                 template<typename...Ts, typename V = View>
                 auto operator()(Ts &&... ts) const

--- a/test/action/CMakeLists.txt
+++ b/test/action/CMakeLists.txt
@@ -10,6 +10,9 @@ add_test(test.act.drop_while act.drop_while)
 add_executable(act.insert insert.cpp)
 add_test(test.act.insert act.insert)
 
+add_executable(act.join join.cpp)
+add_test(test.act.join act.join)
+
 add_executable(act.push_front push_front.cpp)
 add_test(test.act.push_front act.push_front)
 
@@ -33,6 +36,9 @@ add_test(test.act.sort act.sort)
 
 add_executable(act.split split.cpp)
 add_test(test.act.split act.split)
+
+add_executable(act.stable_sort stable_sort.cpp)
+add_test(test.act.stable_sort act.stable_sort)
 
 add_executable(act.stride stride.cpp)
 add_test(test.act.stride act.stride)

--- a/test/action/join.cpp
+++ b/test/action/join.cpp
@@ -1,0 +1,34 @@
+// Range v3 library
+//
+//  Copyright Eric Niebler 2014
+//
+//  Use, modification and distribution is subject to the
+//  Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+
+#include <vector>
+#include <string>
+#include <range/v3/core.hpp>
+#include <range/v3/action/join.hpp>
+#include <range/v3/algorithm/move.hpp>
+#include <range/v3/algorithm/equal.hpp>
+#include <range/v3/view/transform.hpp>
+#include "../simple_test.hpp"
+#include "../test_utils.hpp"
+
+int main()
+{
+    using namespace ranges;
+
+    std::vector<std::string> v {"hello"," ","world"};
+    auto s = v | move | action::join;
+    static_assert(std::is_same<decltype(s), std::string>::value, "");
+    CHECK(s == "hello world");
+
+    auto s2 = v | view::transform(view::all) | action::join;
+    static_assert(std::is_same<decltype(s2), std::vector<char>>::value, "");
+    CHECK(equal(s, s2));
+
+    return ::test_result();
+}

--- a/test/action/stable_sort.cpp
+++ b/test/action/stable_sort.cpp
@@ -1,0 +1,84 @@
+// Range v3 library
+//
+//  Copyright Eric Niebler 2014
+//
+//  Use, modification and distribution is subject to the
+//  Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+
+#include <array>
+#include <random>
+#include <vector>
+#include <range/v3/core.hpp>
+#include <range/v3/view/iota.hpp>
+#include <range/v3/view/stride.hpp>
+#include <range/v3/view/take.hpp>
+#include <range/v3/algorithm/shuffle.hpp>
+#include <range/v3/algorithm/copy.hpp>
+#include <range/v3/algorithm/move.hpp>
+#include <range/v3/algorithm/is_sorted.hpp>
+#include <range/v3/algorithm/equal.hpp>
+#include <range/v3/action/shuffle.hpp>
+#include <range/v3/action/stable_sort.hpp>
+#include "../simple_test.hpp"
+#include "../test_utils.hpp"
+
+void test_bug632()
+{
+	const std::vector<double> scores = { 3.0, 1.0, 2.0 };
+	std::vector<int> indices = { 0, 1, 2 };
+
+	indices |= ranges::action::stable_sort(
+		ranges::less{},
+		[&] (const int &x) { return scores[ (std::size_t)x ]; }
+	);
+
+	::check_equal( indices, {1, 2, 0} );
+}
+
+int main()
+{
+    using namespace ranges;
+    std::mt19937 gen;
+
+    std::vector<int> v = view::ints(0,100);
+    v |= action::shuffle(gen);
+    CHECK(!is_sorted(v));
+
+    auto v2 = v | copy | action::stable_sort;
+    CHECK(size(v2) == size(v));
+    CHECK(is_sorted(v2));
+    CHECK(!is_sorted(v));
+    ::models<concepts::Same>(v, v2);
+
+    v |= action::stable_sort;
+    CHECK(is_sorted(v));
+
+    v |= action::shuffle(gen);
+    CHECK(!is_sorted(v));
+
+    v = v | move | action::stable_sort(std::less<int>());
+    CHECK(is_sorted(v));
+    CHECK(equal(v, v2));
+
+    // Container algorithms can also be called directly
+    // in which case they take and return by reference
+    shuffle(v, gen);
+    CHECK(!is_sorted(v));
+    auto & v3 = action::stable_sort(v);
+    CHECK(is_sorted(v));
+    CHECK(&v3 == &v);
+
+    auto ref=std::ref(v);
+    ref |= action::stable_sort;
+
+    // Can pipe a view to a "container" algorithm.
+    action::stable_sort(v, std::greater<int>());
+    v | view::stride(2) | action::stable_sort;
+    check_equal(view::take(v, 10), {1,98,3,96,5,94,7,92,9,90});
+
+    test_bug632();
+
+    return ::test_result();
+}

